### PR TITLE
Fix Issue 13667 - ICE: stack overflow using self-referencing cast ins…

### PR DESCRIPTION
**Pull request by BorisCarvajal**
…ide recursive alias this method

### CODE REVIEW

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Please target stable, otherwise LGTM

**Comment by BorisCarvajal** [comment]
Can it be merged to master? Another `alias this` fix (#12161) is there and I'm gonna push a little refactoring to replace with a function the same 4 lines of code before every `alias this` resolution.

File: src/dmd/dcast.d [hunk 1]
```diff
@@ -1467,32 +1467,11 @@ Type toStaticArrayType(SliceExp e)
...
-     */
-    result = resolveAliasThis(sc, e);
-    const errors = global.startGagging();
-    result = result.castTo(sc, t);
-    return global.endGagging(errors) ? null : result;
-}
-
 /**************************************
  * Do an explicit cast.
  * Assume that the expression `e` does not have any indirections.
```
Comment (WalterBright) [inline src/dmd/dcast.d:1472]: Document new parameter.

File: src/dmd/dcast.d [hunk 2]
```diff
@@ -1508,6 +1487,32 @@ Expression castTo(Expression e, Scope* sc, Type t)
             this.t = t;
         }
 
+        // Try casting the alias this member. Return the expression if it succeeds, null otherwise.
```
Comment (WalterBright) [inline src/dmd/dcast.d:1490]: What are tob and t1b ?

Comment (WalterBright) [reply src/dmd/dcast.d:1490]: Am I correct that moving tryAliasThisCast here was just a refactoring, and not part of the fix? If so, refactorings (no matter how tempting) should go in separate PRs.

File: src/dmd/opover.d [hunk 3]
```diff
@@ -594,9 +594,12 @@ Expression op_overload(Expression e, Scope* sc, TOK* pop = null)
                     result = result.expressionSemantic(sc);
                     return;
                 }
+                auto t1b = e.e1.type.toBasetype();
                 // Didn't find it. Forward to aliasthis
-                if (ad.aliasthis)
+                if (ad.aliasthis && !(e.att1 && t1b.equivalent(e.att1)))
```
Comment (WalterBright) [inline src/dmd/opover.d:599]: Is this the actual fix?

**Review by WalterBright — DISMISSED** [review]
Please don't mix refactoring with bug fixes. First do one, then the other.

File: src/dmd/dcast.d (hunk 2 above)
Comment (RazvanN7) [reply src/dmd/dcast.d:1490]: It is a refactoring, but it also adds some code.

File: src/dmd/dcast.d [hunk 4]
```diff
@@ -1508,6 +1487,32 @@ Expression castTo(Expression e, Scope* sc, Type t)
...
+                return null;
+
+            AggregateDeclaration toad = isAggregate(tob);
+            if (t1ad == toad || !t1ad.aliasthis)
+                return null;
+
+            if (att1 && t1b.equivalent(att1))
+                return null;
+            else if (!att1 && t1b.checkAliasThisRec())
+                att1 = t1b;
```
Comment (RazvanN7) [inline src/dmd/dcast.d:1504]: @WalterBright  Lines 1501-1504 were not previously in this function.

File: src/dmd/dcast.d (hunk 2 above)
Comment (RazvanN7) [reply src/dmd/dcast.d:1490]: But you are right, the refactoring does make it difficult to spot the added bits.

Comment (BorisCarvajal) [reply src/dmd/dcast.d:1490]: Yep sorry, I've separated refactoring and fix in two commits. It looks more clear now.

File: src/dmd/dcast.d (hunk 4 above)
Comment (BorisCarvajal) [reply src/dmd/dcast.d]: These are the check needed before any `alias this` resolving to avoid infinite recursion.

**Comment by WalterBright** [comment]
Bug fixes and refactorings should be in separate PRs, not just separate commits. The reason is so they can be reviewed independently on their merits, as well as the usual if one introduces a regression, it doesn't implicate the other.

I remember a PR from a few years ago that was shown to be responsible for a regression. The code was around 250 lines of complicated refactoring, and one line of bug fix. It took me a long time to figure that out.

**Comment by BorisCarvajal** [comment]
Ok

**Comment by BorisCarvajal** [comment]
Refactoring -> #12180

**Comment by RazvanN7** [comment]
Rebase and we're good to go.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@WalterBright Do you have any other objections on this PR?

**Pull request merged**